src/transforms/aggregate.rs

```diff
@@ -0,0 +1,440 @@
+use crate::{
+    config::{DataType, GlobalOptions, TransformConfig, TransformDescription},
+    event::{metric, Event, EventMetadata},
+    internal_events::{AggregateEventRecorded, AggregateFlushed, AggregateUpdateFailed},
+    transforms::{TaskTransform, Transform},
+};
+use async_stream::stream;
+use futures::{Stream, StreamExt};
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    pin::Pin,
+    time::Duration,
+};
+
+#[derive(Deserialize, Serialize, Debug, Default, Clone)]
+#[serde(deny_unknown_fields, default)]
+pub struct AggregateConfig {
+    /// The interval between flushes in milliseconds.
+    #[serde(default = "default_interval_ms")]
+    pub interval_ms: u64,
+}
+
+fn default_interval_ms() -> u64 {
+    10 * 1000
+}
+
+inventory::submit! {
+    TransformDescription::new::<AggregateConfig>("aggregate")
+}
+
+impl_generate_config_from_default!(AggregateConfig);
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "aggregate")]
+impl TransformConfig for AggregateConfig {
+    async fn build(&self, _globals: &GlobalOptions) -> crate::Result<Transform> {
+        Aggregate::new(self).map(Transform::task)
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "aggregate"
+    }
+}
+
+type MetricEntry = (metric::MetricData, EventMetadata);
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub struct Aggregate {
+    interval: Duration,
+    map: HashMap<metric::MetricSeries, MetricEntry>,
+}
+
+impl Aggregate {
+    pub fn new(config: &AggregateConfig) -> crate::Result<Self> {
+        Ok(Self {
+            interval: Duration::from_millis(config.interval_ms),
+            map: HashMap::new(),
+        })
+    }
+
+    fn record(&mut self, event: Event) {
+        let (series, data, metadata) = event.into_metric().into_parts();
+
+        match data.kind {
+            metric::MetricKind::Incremental => match self.map.entry(series) {
+                Entry::Occupied(mut entry) => {
+                    let existing = entry.get_mut();
+                    if !existing.0.update(&data) {
+                        emit!(AggregateUpdateFailed);
+                    }
+                    existing.1.merge(metadata);
```

Comment:
In other sinks, when the update fails, we drop the older metric instead of the newer one as is being done here. I think that would be the more sensible behavior, but I am open to your thoughts. Suggestion:
```suggestion
                    if !existing.0.update(&data) {
                        emit!(AggregateUpdateFailed);
                        *existing = (data, metadata);
                    } else {
                        existing.1.merge(metadata);
                    }
```
Note that the assignment causes an implicit drop of the metadata, which is reasonable given that we are also dropping the data as well.

Reply:
Ok. Didn't know the precedent and in a vacuum either option seemed equally bad so I went with the simpler one. Will update now and add in a test for the behavior.

Reply:
Actually @bruceg, it looks like `MetricData.update` succeeds when the kind changes (glad i tried to write tests for that case) so that code is "add"ing `Incremental` and `Absolute` values which doesn't really make sense. I looked into the metrics buffer and I think it's checking for type matches before doing the add. Kind of seems like `update` should do the add or replace logic based on `kind` so that all the users of that method don't have to check and handle that case, but for now I'll rework this to check `kind` first.

Reply:
Ok. Change made in the spirit of your request and the edge cases are now tested and behaving as expected/intended.

Reply:
Yes, it is a new pattern so it wouldn't have been obvious.

Good catch, yes `update` doesn't check the metric kind. It looks like the logic this transform wants and the logic for the metrics buffer is different enough that I'm not clear what the proper semantics should be. I'll have to give that more thought.

Reply:
If nothing else it could fail the update, do nothing and return false. Assuming something elsewhere isn't expecting/wanting it to add things with non-matching `kind`.

Reply:
Yeah, I was thinking more along the lines of `update` possibly taking the new data by value instead of reference and returning it back in the case of an error as  `Result<(), MetricData>`, to avoid needing to clone data inside update. On the other hand, the metrics buffer wants to be able to do add incrementals to both kinds, while this transform wants to replace for the case of different kinds, so that's hard to shoehorn into one method.

Reply:
Guess it's not clear to me how adding `Absolute`s makes sense given they're intended to be the value as-is, e.g. if you're going 55mph at t0 and 57mpg at t1, 112mph isn't a meaningful number. It seems like `kind` and `Counter`/`Guage` kind of overlap semantic wise and can result in internally conflicting states.

Anyway, not really related to this PR so I can let it go :grin:

Reply:
No, you can't add absolutes, but you can *update* an absolute to a new absolute value. You can *add* an incremental to either kind as well. Agreed it's not related to this PR.